### PR TITLE
[NativeAOT-LLVM] Introduce a shadow stack allocation phase

### PR DESCRIPTION
**Pull request by SingleAccretion**
For now, just move code to it from lowering.

This will fix the recently introduced GC hole with interface dispatch.

Diffs are regressions caused by the fix:
```
Total bytes of base: 3414056
Total bytes of diff: 3431153
Total bytes of delta: 17097 (0.50% % of base)
Average relative delta: 8.01%
    diff is a regression
    average relative diff is a regression

Top method regressions (percentages):
          56 (58.33% of base) : 1390.dasm - S_P_CoreLib_System_Convert__ToChar
          56 (58.33% of base) : 1389.dasm - S_P_CoreLib_System_Convert__ToBoolean
          56 (58.33% of base) : 1461.dasm - S_P_CoreLib_System_Convert__ToSByte
          56 (58.33% of base) : 1462.dasm - S_P_CoreLib_System_Convert__ToByte
          56 (58.33% of base) : 1463.dasm - S_P_CoreLib_System_Convert__ToInt16
          56 (58.33% of base) : 1464.dasm - S_P_CoreLib_System_Convert__ToUInt16
          56 (58.33% of base) : 1466.dasm - S_P_CoreLib_System_Convert__ToUInt32
          56 (58.33% of base) : 1465.dasm - S_P_CoreLib_System_Convert__ToInt32
          50 (52.08% of base) : 1468.dasm - S_P_CoreLib_System_Convert__ToUInt64
          50 (52.08% of base) : 1467.dasm - S_P_CoreLib_System_Convert__ToInt64
          50 (50.51% of base) : 1469.dasm - S_P_CoreLib_System_Convert__ToSingle
          12 (50.00% of base) : 1117.dasm - S_P_CoreLib_System_Reflection_Runtime_General_NativeFormatMetadataReaderExtensions__ToExpectedNamespaceReferenceHandle
          50 (48.54% of base) : 1470.dasm - S_P_CoreLib_System_Convert__ToDouble
          30 (33.71% of base) : 1272.dasm - S_P_CoreLib_System_Type__GetTypeFromHandle
          29 (32.95% of base) : 1551.dasm - S_P_CoreLib_System_Globalization_TextInfo__NeedsTurkishCasing
          24 (30.38% of base) : 1586.dasm - S_P_CoreLib_Internal_Runtime_CompilerHelpers_LdTokenHelpers__GetRuntimeType
          89 (30.27% of base) : 1224.dasm - S_P_CoreLib_System_Text_StringBuilder_AppendInterpolatedStringHandler__AppendFormatted_4
          24 (29.63% of base) : 1009.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_Callbacks__CompareMethodSignatures
          24 (29.63% of base) : 1010.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_Callbacks__GetThreadStaticGCDescForDynamicType
          24 (29.63% of base) : 1008.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_Callbacks__ResolveGenericVirtualMethodTarget

Top method improvements (percentages):
         -64 (-13.91% of base) : 1054.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_MetadataNameExtensions__GetFullName_6
         -14 (-9.59% of base) : 1735.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_MetadataNameExtensions__GetFullName_0
         -14 (-9.59% of base) : 1654.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_MetadataNameExtensions__GetFullName_3
         -14 (-9.59% of base) : 1655.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_MetadataNameExtensions__GetFullName_1
         -35 (-8.95% of base) : 1409.dasm - S_P_TypeLoader_Internal_Metadata_NativeFormat_MetadataTypeHashingAlgorithms__ComputeHashCode
         -54 (-7.85% of base) : 1292.dasm - S_P_CoreLib_System_DateTimeFormat__ExpandStandardFormatToCustomPattern
          -3 (-4.55% of base) : 1241.dasm - S_P_CoreLib_System_Reflection_Runtime_General_Helpers__GetRawDefaultValue$F1_Finally
          -3 (-4.55% of base) : 1565.dasm - S_P_TypeLoader_Internal_Runtime_TypeLoader_TypeLoaderEnvironment__RegisterDynamicGenericTypesAndMethods$F1_Fault
          -3 (-4.55% of base) : 1573.dasm - S_P_CoreLib_System_Reflection_Runtime_General_Helpers__GetRawConstant$F1_Finally
          -3 (-4.55% of base) : 1239.dasm - S_P_CoreLib_System_Reflection_Runtime_General_Helpers__GetDefaultValue$F1_Finally
          -3 (-4.55% of base) : 1091.dasm - S_P_CoreLib_System_Collections_Generic_LowLevelList_1<System___Canon>__InsertRange$F1_Fault
          -3 (-4.55% of base) : 1342.dasm - S_P_TypeLoader_Internal_TypeSystem_TypeDesc__get_UnderlyingType$F1_Finally
          -3 (-4.55% of base) : 1619.dasm - S_P_CoreLib_System_Reflection_Runtime_BindingFlagSupport_Shared__GetImplicitlyOverriddenBaseClassMember<System___Canon>$F1_Finally
          -3 (-4.55% of base) : 1075.dasm - S_P_CoreLib_System_Reflection_Runtime_BindingFlagSupport_QueriedMemberList_1<System___Canon>__Create$F1_Fault
         -72 (-4.42% of base) : 1045.dasm - S_P_StackTraceMetadata_Internal_StackTraceMetadata_MethodNameFormatter__EmitTypeName
         -20 (-4.12% of base) : 1116.dasm - S_P_CoreLib_Microsoft_Win32_SafeHandles_SafeFileHandle__Open
         -10 (-4.07% of base) : 1209.dasm - String__Create<IntPtr>
         -10 (-3.26% of base) : 1186.dasm - S_P_CoreLib_System_Reflection_Runtime_TypeInfos_RuntimeTypeInfo_TypeComponentsCache__GetQueriedMembers_0<System___Canon>
         -38 (-3.10% of base) : 1346.dasm - S_P_Reflection_Execution_Internal_Reflection_Execution_ExecutionEnvironmentImplementation__TryGetFieldAccessor
          -3 (-2.94% of base) : 1113.dasm - S_P_CoreLib_System_Reflection_Runtime_CustomAttributes_RuntimeCustomAttributeData__WrapInCustomAttributeTypedArgument$F1_Fault
```

I am also planning to add a stress mode (in a separate change) that would help stop such issues from slipping through in the future.

### CODE REVIEW

File: src/coreclr/jit/llvm.h [hunk 1]
```diff
@@ -350,19 +342,38 @@ class Llvm
...
     GenTree* insertShadowStackAddr(GenTree* insertBefore, ssize_t offset, unsigned shadowStackLclNum);
 
+    unsigned getCatchArgOffset() const;
+
+    // ================================================================================================================
+    // |                                           Shadow stack allocation                                            |
+    // ================================================================================================================
+public:
+    void Allocate();
+
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.h:354]: I expect that this code structure is more or less temporary and LSSA proper will be put into its own class.

File: src/coreclr/jit/llvm.h [hunk 2]
```diff
@@ -350,19 +342,38 @@ class Llvm
...
+    void allocDissolvePromotedLocals();
+    void allocAssignShadowFrameOffsets(std::vector<unsigned>& shadowFrameLocals);
+
+    void allocLowerAndInsertProlog();
+    void allocInitializeLocalInProlog(unsigned lclNum, GenTree* value);
+
+    void allocRewriteShadowFrameReferences();
+    void allocRewriteLocal(GenTreeLclVarCommon* lclNode);
+    void allocRewriteCall(GenTreeCall* call);
+
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvm.h:368]: I expect that this code structure is more or less temporary and LSSA proper will be put into its own class.

**Comment by SingleAccretion** [comment]
@dotnet/nativeaot-llvm

**Review by yowl — COMMENTED** [review]
Can you point out for my benefit where the regression and the GC hole are in the changes?  Thanks

**Comment by SingleAccretion** [comment]
@yowl The GC hole was that for an interface call:
```
gcArg = ...

call(gcArg)
```
Inserting the resolver created a new safepoint across which GC arguments to the call would be live:
```
gcArg = ...
ResolveTarget(...) ; Could trigger GC and relocate 'gcArg'
call(gcArg)
```
The fix is that the pass which makes sure this does not happen (`SpillTempsLiveAcrossSafePoints`) is moved after lowering, so any vulnerable arguments will be spilled to the shadow stack as appropriate. One can notice that this means pretty much all GC arguments in interface calls have to be spilled, hence the regressions.

**Review by yowl — APPROVED** [review]
Thank you.  Looks good.

**Pull request merged**